crates/ty_python_semantic/resources/mdtest/type_qualifiers/initvar.md

```diff
@@ -0,0 +1,124 @@
+# `dataclasses.InitVar`
+
+From the Python documentation on [`dataclasses.InitVar`]:
+
+If a field is an `InitVar`, it is considered a pseudo-field called an init-only field. As it is not
+a true field, it is not returned by the module-level `fields()` function. Init-only fields are added
+as parameters to the generated `__init__()` method, and are passed to the optional `__post_init__()`
+method. They are not otherwise used by dataclasses.
+
+## Basic
+
+Consider the following dataclass example where the `db` attribute is annotated with `InitVar`:
+
+```py
+from dataclasses import InitVar, dataclass
+
+class Database: ...
+
+@dataclass(order=True)
+class Person:
+    db: InitVar[Database]
+
+    name: str
+    age: int
+```
+
+We can see in the signature if `__init__`, that `db` is included as an argument:
+
+```py
+reveal_type(Person.__init__)  # revealed: (self: Person, db: Database, name: str, age: int) -> None
+```
+
+However, when we create an instance of this dataclass, the `db` attribute is not accessible:
+
+```py
+db = Database()
+alice = Person(db, "Alice", 30)
+
+alice.db  # error: [unresolved-attribute]
+```
+
+The `db` attribute is also not accessible on the class itself:
+
+```py
+Person.db  # error: [unresolved-attribute]
+```
+
+Other fields can still be accessed normally:
+
+```py
+reveal_type(alice.name)  # revealed: str
+reveal_type(alice.age)  # revealed: int
+```
+
+## `InitVar` wit default value
+
+An `InitVar` can also have a default value
+
+```py
+from dataclasses import InitVar, dataclass
+
+@dataclass
+class Person:
+    name: str
+    age: int
+
+    metadata: InitVar[str] = "default"
+
+reveal_type(Person.__init__)  # revealed: (self: Person, name: str, age: int, metadata: str = Literal["default"]) -> None
+
+alice = Person("Alice", 30)
+bob = Person("Bob", 25, "custom metadata")
+```
```

Comment:
I don't think it's hugely important for us to support this, but note that if you provide a default value for an `InitVar` then it _is_ accessible from both classes and instances:

```pycon
>>> from dataclasses import InitVar, dataclass
... 
... @dataclass
... class Person:
...     name: str
...     age: int
... 
...     metadata: InitVar[str] = "default"
...     
>>> Person.metadata
'default'
>>> p = Person("alice", 42)
>>> p.metadata
'default'
```

but it's still not a _true_ dataclass field: it's not considered for equality, etc.:

```pycon
>>> p = Person("alice", 42)
>>> p2 = Person("alice", 42)
>>> p2.metadata = 'whatever'
>>> p == p2
True
```

Reply:
> I don't think it's hugely important for us to support this, but note that if you provide a default value for an `InitVar` then it _is_ accessible from both classes and instances:

Oh! Maybe that is important to reflect, see the ecosystem changes in bokeh.